packages/upload-api/src/ucan/conclude.js

```diff
@@ -1,106 +1,33 @@
-      if (httpPutTaskGetRes.error.name === 'RecordNotFound') {
-        return {
-          error: new ReferencedInvocationNotFound(ranInvocation.link()),
-        }
-      }
-      return httpPutTaskGetRes
-    }
-
-    // Store receipt
-    const receiptPutRes = await receiptsStorage.put(receipt)
-    if (receiptPutRes.error) {
-      return {
-        error: receiptPutRes.error,
-      }
-    }
-
-    // THIS IS A TEMPORARY HACK
-    // Schedule `blob/accept` if there is a `http/put` capabilities
-    // inside the invocation that this receipt comes from
-    const scheduleRes = await Promise.all(
-      httpPutTaskGetRes.ok.capabilities
-        // Go through invocation tasks and get all `http/put`
-        .filter((cap) => cap.can === HTTP.put.can)
-        // @ts-expect-error body exists in `http/put` but unknown type here
-        .map(async (/** @type {API.HTTPPut} */ cap) => {
-          // Get triggering task (blob/allocate) by checking blocking task from `url`
-          /** @type {API.UnknownLink} */
-          // @ts-expect-error ts does not know how to get this
-          const [, blobAllocateTaskCid] = cap.nb.url['ucan/await']
-          const blobAllocateTaskGet = await tasksStorage.get(
-            blobAllocateTaskCid
-          )
-          if (blobAllocateTaskGet.error) {
-            return blobAllocateTaskGet
-          }
-
-          /** @type {API.BlobAllocate} */
-          // @ts-expect-error ts does not know how to get this
-          const allocateCapability = blobAllocateTaskGet.ok.capabilities.find(
-            // @ts-expect-error ts does not know how to get this
-            (/** @type {API.BlobAllocate} */ allocateCap) =>
-              equals(allocateCap.nb.blob.digest, cap.nb.body.digest) &&
-              allocateCap.can === W3sBlob.allocate.can
-          )
-
-          const blobAccept = await W3sBlob.accept
-            .invoke({
-              issuer: id,
-              audience: id,
-              with: id.did(),
-              nb: {
-                blob: cap.nb.body,
-                space: allocateCapability.nb.space,
-                _put: {
-                  'ucan/await': ['.out.ok', ranInvocation.link()],
-                },
-              },
-              // Expiry is set to `Infinity` so that CID will come out the same
-              // as returned in effect on `blob/add`
-              expiration: Infinity,
-            })
-            .delegate()
-
-          return tasksScheduler.schedule(blobAccept)
-        })
-    )
-
-    const scheduleErrors = scheduleRes.filter((res) => res.error)
-    if (scheduleErrors.length && scheduleErrors[0].error) {
-      return {
-        error: scheduleErrors[0].error,
-      }
+    // First of all that we create a message and save it to agent store, that
+    // way even if invocation occurs later we will have receipt already.
+    const message = await Message.build({ receipts: [receipt] })
+    const save = await context.agentStore.messages.write(message)
```

Comment:
is the goal here to save the message and all the indexes to it like w3infra does today?

Reply:
Conceptually I'm thinking of `AgentStore` as ucanto agent cache. Implementation of such cache can choose cache invalidation strategy best suited for the use case. E.g. w3infra could invalidate receipts for invocations it does not care about (has seen) or it could store them all in issuers space, if there is one or discard otherwise. It could also just store everything and worry about GC afterwards perhaps utilizing something like LRU evection strategy.

Bigger picture aside, yes I was intending to just store receipts in the receipt store we have setup. I realize it could end up storing things we don't care about, but that is also the case with delegations store and bunch of other things. I don't worry about it much because there is very little to be gained by sending us pointless receipts, but perhaps not storing receipts for invocations we have not seen is a better default, which I suspect is you're concern here.

Simply storing and not worrying we get following

1. Simpler code that is easier to maintain.
2. Support for out of order delivery (that is we may see receipt first and invocation after)

What do you think ?

Reply:
What I was meaning was more, if this first iteration of the AgentStore would do exactly same as we do in w3infra and store things into `workflow`, `task` and `invocation` buckets.

I believe in the long run, we should do what we talked about and store this at the appropriate space, but now would likely be good to not think about a new way of storing and reading things, as a migration and so on would also need to be done

Reply:
in other words, if the store initial impl would so something like what happens in https://github.com/w3s-project/w3infra/blob/main/upload-api/functions/ucan-invocation-router.js#L318

Reply:
> in other words, if the store initial impl would so something like what happens in https://github.com/w3s-project/w3infra/blob/main/upload-api/functions/ucan-invocation-router.js#L318

It is exactly what I defined the interface from with intention to do exactly that